### PR TITLE
feat: [CDS-47984]: Ecs Run Task ARN

### PR DESCRIPTION
**Pull request by alluvamsi-harness**
## Describe your changes
To support Task Definition as ARN in Run Task Step of ECS NG
https://docs.google.com/spreadsheets/d/10thbPOV0enoLZXTr6MGXiORestUN06_ONmxbbtxZESM/edit#gid=0
## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/41647)

### CODE REVIEW

**Comment by alluvamsi-harness** [comment]
trigger feature-build

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41647)* status: 12 file(s) to review. 0 of 2 approvals obtained.

File: 930-delegate-tasks/src/main/java/io/harness/delegate/ecs/EcsRunTaskArnCommandTaskHandler.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+  @Inject private EcsTaskHelperBase ecsTaskHelperBase;
+  @Inject private EcsCommandTaskNGHelper ecsCommandTaskHelper;
+
+  private EcsInfraConfig ecsInfraConfig;
+  private long timeoutInMillis;
+  @Override
+  protected EcsCommandResponse executeTaskInternal(EcsCommandRequest ecsCommandRequest,
+      ILogStreamingTaskClient iLogStreamingTaskClient, CommandUnitsProgress commandUnitsProgress) throws Exception {
+    if (!(ecsCommandRequest instanceof EcsRunTaskArnRequest)) {
+      throw new InvalidArgumentsException(Pair.of("ecsCommandRequest", "Must be instance of EcsRollingDeployRequest"));
```
Comment (sainath-b) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/ecs/EcsRunTaskArnCommandTaskHandler.java:58]: exception message need to be changed to EcsRunTaskArnRequest

File: 125-cd-nextgen/src/main/java/io/harness/cdng/ecs/EcsRunTaskStep.java [hunk 2]
```diff
@@ -141,22 +142,40 @@ public TaskChainResponse executeEcsTask(Ambiance ambiance, StepElementParameters
...
-            .ecsRunTaskRequestDefinitionManifestContent(
-                ecsStepExecutorParams.getEcsRunTaskRequestDefinitionManifestContent())
-            .skipSteadyStateCheck(ecsRunTaskStepParameters.getSkipSteadyStateCheck().getValue())
-            .build();
-
-    return ecsStepCommonHelper.queueEcsTask(
-        stepElementParameters, ecsRunTaskRequest, ambiance, executionPassThroughData, true);
+    if (ecsRunTaskStepParameters.getTaskDefinitionArn() == null
+        || ecsRunTaskStepParameters.getTaskDefinitionArn().getValue() == null) {
+      EcsRunTaskRequest ecsRunTaskRequest =
```
Comment (PragyeshMishra01) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/ecs/EcsRunTaskStep.java:147]: what will happen if taskDefinitionArn is expression?

File: 930-delegate-tasks/src/main/java/io/harness/delegate/ecs/EcsRunTaskArnCommandTaskHandler.java (hunk 1 above)
Comment (alluvamsi-harness) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/ecs/EcsRunTaskArnCommandTaskHandler.java]: Done

File: 125-cd-nextgen/src/main/java/io/harness/cdng/ecs/EcsRunTaskStep.java (hunk 2 above)
Comment (alluvamsi-harness) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/ecs/EcsRunTaskStep.java:147]: added this cdExpressionResolver.updateExpressions(ambiance, stepElementParameters);
expression gets resolved and value will be replaced 
Tested expressions and secretes working fine now updated in test plan as well

File: 125-cd-nextgen/src/main/java/io/harness/cdng/ecs/EcsStepCommonHelper.java [hunk 3]
```diff
@@ -448,20 +452,35 @@ public List<ManifestOutcome> getEcsManifestOutcome(
...
+            || ecsRunTaskStepParameters.getTaskDefinitionArn().getValue() == null)) {
       String errorMessage = "ECS Task Definition is empty in ECS Run Task Step";
       throw new InvalidRequestException(errorMessage);
     }
 
+    if ((ecsRunTaskStepParameters.getTaskDefinition() != null
+            && ecsRunTaskStepParameters.getTaskDefinition().getValue() != null)
+        && (ecsRunTaskStepParameters.getTaskDefinitionArn() != null
+            && ecsRunTaskStepParameters.getTaskDefinitionArn().getValue() != null)) {
+      String errorMessage = "Configure only one Task Definition";
```
Comment (sainath-b) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/ecs/EcsStepCommonHelper.java:467]: Error message could be better here. We can say 'Both Task Definition, Task Defintion Arn are configured. Only one of them is expected.'

Comment (alluvamsi-harness) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/ecs/EcsStepCommonHelper.java:467]: Ok changing it

**Review by PragyeshMishra01 — CHANGES_REQUESTED** [review]

File: 930-delegate-tasks/src/main/java/io/harness/delegate/ecs/EcsRunTaskArnCommandTaskHandler.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+    String ecsTaskDefinition = ecsRunTaskArnRequest.getEcsTaskDefinition();
+    runTaskLogCallback.saveExecutionLog(format("TaskDefinition: %s %n", ecsTaskDefinition), LogLevel.INFO);
+    DescribeTaskDefinitionResponse describeTaskDefinitionResponse =
+        ecsCommandTaskHelper.validateEcsTaskDefinition(ecsTaskDefinition, ecsInfraConfig, runTaskLogCallback);
+
+    TaskDefinition taskDefinition = describeTaskDefinitionResponse.taskDefinition();
+
+    return ecsCommandTaskHelper.getEcsRunTaskResponse(taskDefinition,
+        ecsRunTaskArnRequest.getEcsRunTaskRequestDefinitionManifestContent(),
```
[the remark this answers is not in the document]
Comment (alluvamsi-harness) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/ecs/EcsRunTaskArnCommandTaskHandler.java:66]: Sure

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/ecs/EcsCommandTaskNGHelper.java [hunk 5]
```diff
@@ -1266,6 +1281,90 @@ public boolean isServiceDraining(Service service) {
...
+
+    RunTaskResponse runTaskResponse =
+        ecsCommandTaskHelper.runTask(runTaskRequest, ecsInfraConfig.getAwsConnectorDTO(), ecsInfraConfig.getRegion());
+
+    runTaskLogCallback.saveExecutionLog(format("%d Tasks were triggered successfully and %d failures were received.",
+                                            runTaskResponse.tasks().size(), runTaskResponse.failures().size()),
+        LogLevel.INFO);
+
+    runTaskResponse.tasks().forEach(
+        t -> runTaskLogCallback.saveExecutionLog(format("Task => %s succeeded", t.taskArn())));
```
[the remark this answers is not in the document]
Comment (alluvamsi-harness) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/ecs/EcsCommandTaskNGHelper.java:1309]: task() is always initialized so even passing an empty list will not throw NPE

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/ecs/EcsCommandTaskNGHelper.java [hunk 6]
```diff
@@ -1266,6 +1281,90 @@ public boolean isServiceDraining(Service service) {
...
+
+    runTaskResponse.failures().forEach(f -> {
+      runTaskLogCallback.saveExecutionLog(
+          format("%s failed with reason => %s \nDetails: %s", f.arn(), f.reason(), f.detail()), LogLevel.ERROR,
+          CommandExecutionStatus.FAILURE);
+    });
```
[the remark this answers is not in the document]
Comment (alluvamsi-harness) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/ecs/EcsCommandTaskNGHelper.java:1315]: failures() is always initialized so even passing an empty list will not throw NPE, failures will never be Null

File: 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/ecs/request/EcsRunTaskArnRequest.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+  EcsCommandTypeNG ecsCommandType;
+  String commandName;
+  CommandUnitsProgress commandUnitsProgress;
+  boolean skipSteadyStateCheck;
+  @NonFinal @Expression(ALLOW_SECRETS) EcsInfraConfig ecsInfraConfig;
+  @NonFinal @Expression(ALLOW_SECRETS) Integer timeoutIntervalInMin;
+
+  @Expression(ALLOW_SECRETS) String ecsTaskDefinition;
+  @Expression(ALLOW_SECRETS) String ecsRunTaskRequestDefinitionManifestContent;
+}
```
[the remark this answers is not in the document]
Comment (alluvamsi-harness) [reply 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/ecs/request/EcsRunTaskArnRequest.java:38]: Ok adding it

File: 930-delegate-tasks/src/main/java/io/harness/delegate/ecs/EcsRunTaskArnCommandTaskHandler.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (alluvamsi-harness) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/ecs/EcsRunTaskArnCommandTaskHandler.java:66]: Sure

[the remark this answers is not in the document]
Comment (alluvamsi-harness) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/ecs/EcsRunTaskArnCommandTaskHandler.java:66]: Inside that function error handling is handled in different scenarios which will be caught in EcsDelegateTaskHelper and thrown as TaskNGDataException

[the remark this answers is not in the document]
Comment (alluvamsi-harness) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/ecs/EcsRunTaskArnCommandTaskHandler.java:66]: added .

**Comment by alluvamsi-harness** [comment]
trigger smartchecks

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/ecs/EcsCommandTaskNGHelper.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (alluvamsi-harness) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/ecs/EcsCommandTaskNGHelper.java]: added

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/ecs/EcsCommandTaskNGHelper.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (alluvamsi-harness) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/ecs/EcsCommandTaskNGHelper.java]: added

**Comment by alluvamsi-harness** [comment]
trigger checkpermission

**Review by PragyeshMishra01 — APPROVED** [review]

**Comment by alluvamsi-harness** [comment]
trigger checkpermission

**Comment by alluvamsi-harness** [comment]
trigger checkpermission

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 3 of 10 files at r1, 1 of 2 files at r3, 5 of 16 files at r4, 2 of 11 files at r5, 1 of 5 files at r6, 6 of 8 files at r7, 4 of 4 files at r8, 5 of 5 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41647)* status: 25 file(s) to review. 7. unresolved discussions.

**Review by anilchowdhury-harness — APPROVED** [review]
:lgtm:

> Reviewed 3 of 10 files at r1, 1 of 2 files at r3, 5 of 16 files at r4, 2 of 11 files at r5, 1 of 5 files at r6, 6 of 8 files at r7, 4 of 4 files at r8, 5 of 5 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41647#-:-NJtjGQ8DGEPywowHWkq:bnfp4nl)* status: 7. unresolved discussions.

**Review by PragyeshMishra01 — APPROVED** [review]
> Reviewed 3 of 10 files at r1, 1 of 2 files at r3, 5 of 16 files at r4, 2 of 11 files at r5, 1 of 5 files at r6, 6 of 8 files at r7, 4 of 4 files at r8, 5 of 5 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41647)* status: 2. unresolved discussions.

**Review by PragyeshMishra01 — APPROVED** [review]
> Reviewed 1 of 4 files at r8.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41647)* status: 2. unresolved discussions.

**Review by PragyeshMishra01 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41647)* status: 2. unresolved discussions.

**Review by alluvamsi-harness — COMMENTED** [review]
> Reviewed 2 of 4 files at r8.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41647)* status: 2. unresolved discussions.

**Review by mdmilic — APPROVED** [review]
approve delegate code owned changes

> Reviewed 4 of 8 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41647#-:-NJu5lY05OX1RNu7cf8l:b-ecdtpk)* status: 2. unresolved discussions.

**Review by sainath-b — APPROVED** [review]

**Review by sainath-b — COMMENTED** [review]
> Reviewed 3 of 10 files at r1, 1 of 2 files at r3, 5 of 16 files at r4, 2 of 11 files at r5, 1 of 5 files at r6, 6 of 8 files at r7, 4 of 4 files at r8, 5 of 5 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41647)* status: 2. unresolved discussions.

**Review by sainath-b — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41647)* status: :shipit: complete! ✓

**Comment by sainath-b** [comment]
@alluvamsi-harness 
Approved this PR. But as discussed, please validate ecs task definition improvment as a part of another PR

**Pull request merged**